src/dupin.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <Eigen/Dense>
+#include <algorithm>
+#include <iostream>
+#include <limits>
+#include <omp.h>
+#include <unordered_map>
+#include <vector>
+
+// DynamicProgramming class for dynamic programming based segmentation.
+class DynamicProgramming { // change name to Dynamic Programming
+private:
+  class upper_triangular_cost_matrix {
+  private:
+    std::vector<double> matrix;
+    int size;
+
+    int index(int i, int j) const {
+      return i * (2 * size - i + 1) / 2 + (j - i);
+    }
+
+  public:
+    upper_triangular_cost_matrix() : size(0) {}
+
+    void initialize(int n) {
+      size = n;
+      matrix.resize(n * (n + 1) / 2, 0.0);
+    }
+
+    double &operator()(int i, int j) { return matrix[index(i, j)]; }
+
+    int getSize() const { return size; }
+  };
+  upper_triangular_cost_matrix cost_matrix;
+  // Struct for memoization key, combining start, end, and number of
+  // breakpoints.
+  struct MemoKey {
+    int start;
+    int end;
+    int num_bkps;
+
+    // Comparison operator for MemoKey.
+    bool operator==(const MemoKey &other) const {
+      return start == other.start && end == other.end &&
+             num_bkps == other.num_bkps;
+    }
+  };
+
+  // Custom XOR-bit hash function for MemoKey, avoids clustering of data in
+  // unordered map to improve efficiency.
+  struct MemoKeyHash { // test without hash function
+    std::size_t operator()(const MemoKey &key) const {
+      return ((std::hash<int>()(key.start) ^
+               (std::hash<int>()(key.end) << 1)) >>
+              1) ^
+             std::hash<int>()(key.num_bkps);
+    }
+  };
+
+  // Memoization map to store the cost and partition for given parameters.
+  std::unordered_map<MemoKey, std::pair<double, std::vector<int>>, MemoKeyHash>
+      memo;
+
+  int num_bkps;          // Number of breakpoints to detect.
+  int num_parameters;    // Number of features in the dataset.

```

Comment:
This should be known from the passed matrix no?

Reply:
Some of this was just for testing/setting back in ye old days of input.txt. I can just make it use data.size now

Reply:
Still to be done.

Reply:
If we keep this we should be more elaborate and use `num_features` and `num_breakpoints`.

Reply:
Also, if possible, expand on the variable names throughout. It will help future developers work with the code.